neo/Storage/Implementation/LevelDB/LevelDBClassMethods.py

```diff
@@ -0,0 +1,235 @@
+import plyvel
+import threading
+
+from contextlib import contextmanager
+
+from neo.Core.Blockchain import Blockchain
+from neo.Storage.Common.DBPrefix import DBPrefix
+from neo.Storage.Interface.DBInterface import DBProperties
+from neo.logging import log_manager
+
+
+logger = log_manager.getLogger()
+
+"""
+Description:
+    All the methods are used within the dynamically generated class from the
+    database factory.
+Usage:
+    For a new database implementation all methods defined in
+    AbstractDBImplementation have to be implemented.
+"""
+
+# name of the init method for a generic database class
+_init_method = '_db_init'
+
```

Comment:
I had to look in the code to understand what the follow section does, which signals that we can improve the documentation to make it more self-explanatory. While saying that I now realise that `AbstractDBImplementation` doesn't cover all the necessary requirements to support an alternative backend. We actually also need to specify a specific set of variables like `_init_method` such that it can be picked up by 
```
        # add __init__ method
        attributes['__init__'] = attributes.pop(functions._init_method)
```
from the factory. 

I should have noticed this sooner but I think the current factory is limited in the way that it needs modification to support new backends. This can be seen by looking at your `Rocksdb` update
https://github.com/merl111/neo-python/blob/54970a06316f54ebbca86d4bb573404ae197ddbe/neo/Storage/Implementation/DBFactory.py#L115-L116

So why does this matter? It means that if somebody creates an alternative backend then there are 2 options for using it
1) it remains user maintained and for every `neo-python` update they need to patch it back in for usage
2) it becomes part of `neo-python` and maintenance (partially) becomes our responsibility. 

I'm thinking 2 things:
1) I'd like maintenance of alternative DBs to not be `neo-python`'s responsibility. 
2) I'd like to see if we can actually inherit from the `AbstractDBImplementation` and add _all_ required items/methods that need to be overridden in one place.

The maintenance thing can be resolved by making the DB class to be used to be loaded from the settings file. Similar to Django's settings/config. We actually do this for the RPC and REST server. 

https://github.com/CityOfZion/neo-python/blob/fe90f62e123d720d4281c79af0598d9df9e776fb/neo/data/protocol.privnet.json#L44

https://github.com/CityOfZion/neo-python/blob/fe90f62e123d720d4281c79af0598d9df9e776fb/neo/bin/api_server.py#L282

If you have any questions regarding anything above, hit me up on Discord.

Reply:
> I'd like maintenance of alternative DBs to not be neo-python's responsibility.

Ok, my first impression was that those different backend implementation should be part of neo-python, but if that's not the case my approach might not be perfect. I have changed a few things, `DBFactory` does not need any changes anymore if one implements a different backend and I implemented your suggestion regarding to load the class from the settings file. Please check last commit.

> I'd like to see if we can actually inherit from the AbstractDBImplementation and add all required items/methods that need to be overridden in one place.

I also changed the LevelDB implementation a bit, so, if I am not missing anything, all required methods that need to be overridden should be in the `AbstractDBImplementation`.